pest/src/error.rs

```diff
@@ -5,115 +5,266 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-use super::Input;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+
+use super::inputs::{Input, Position, Span};
 
 /// An `enum` which defines possible errors.
-#[derive(Clone, Debug, Eq, Hash, PartialEq)]
-pub enum Error<Rule> {
+#[derive(Debug)]
+pub enum Error<R, I: Input> {
     /// Generated parsing error with expected and unexpected `Rule`s and a position
-    ParsingError(Vec<Rule>, Vec<Rule>, usize),
+    ParsingError {
+        /// Positive attempts
+        positives: Vec<R>,
+        /// Negative attempts
+        negatives: Vec<R>,
+        /// Deepest position of attempts
+        pos: Position<I>
+    },
     /// Custom error with a message and a position
-    CustomErrorPos(String, usize),
+    CustomErrorPos {
+        /// Short explanation
+        message: String,
+        /// Error `Position` for formatting
+        pos: Position<I>
+    },
     /// Custom error with a message and a span defined by a start and end position
-    CustomErrorSpan(String, usize, usize)
+    CustomErrorSpan {
+        /// Short explanation
+        message: String,
+        /// Error `Span` for formatting
+        span: Span<I>
+    }
 }
 
-impl<Rule> Error<Rule> {
-    /// Formats an `Error` according to an `Input`.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// # use pest::{Error, StringInput};
-    /// let input = StringInput::new("ab\ncd\nef");
-    /// let error: Error<()> = Error::CustomErrorPos("error: big one".to_owned(), 4);
-    ///
-    /// assert_eq!(error.format(&input), vec![
-    ///     " --> 2:2",
-    ///     "  |",
-    ///     "2 | cd",
-    ///     "  |  ^---",
-    ///     "  |",
-    ///     "  = error: big one"
-    /// ].join("\n"));
-    /// ```
-    pub fn format<I: Input>(&self, input: &I) -> String {
-        self.format_option(input, None)
+fn message<R: fmt::Debug, I: Input>(error: &Error<R, I>) -> String {
+    match *error {
+        Error::ParsingError { ref positives, ref negatives, .. } => {
+            match (negatives.is_empty(), positives.is_empty()) {
+                (false, false) => {
+                    format!(
+                        "unexpected {}; expected {}",
+                        enumerate(negatives),
+                        enumerate(positives)
+                    )
+                }
+                (false, true) => format!("unexpected {}", enumerate(negatives)),
+                (true, false) => format!("expected {}", enumerate(positives)),
+                (true, true) => "inexplicit parsing error".to_owned()
```

Comment:
I guess "unknown" is a bit clearer as a general error. The reason I used "inexplicit" was to encourage shaping the grammar in a way that would make the error explicit.